typed-racket-lib/typed-racket/typecheck/tc-expr-unit.rkt

```diff
@@ -387,22 +387,44 @@
      (match (and expected (resolve (intersect expected (-pair Univ Univ))))
        [(Pair: car-t cdr-t) (-pair (find-stx-type car car-t) (find-stx-type cdr cdr-t))]
        [_ (-pair (find-stx-type car) (find-stx-type cdr))])]
-    [(vector xs ...)
-     (match (and expected (resolve (intersect expected -VectorTop)))
-       [(Vector: t)
-        (make-Vector
-         (check-below
-          (apply Un
-                 (for/list ([x (in-list xs)])
-                   (find-stx-type x t)))
-          t))]
-       [(HeterogeneousVector: ts)
-        (make-HeterogeneousVector
-         (for/list ([x (in-list xs)]
-                    [t (in-list/rest ts #f)])
-           (cond-check-below (find-stx-type x t) t)))]
-       [_ (make-HeterogeneousVector (for/list ([x (in-list xs)])
-                                      (generalize (find-stx-type x #f))))])]
+    [(and (vector xs ...) v)
```

Comment:
Sam points out: `datum->syntax` is documented to turn mutable vectors into immutable vectors. So that explains the behavior in `typecheck-tests.rkt`